banderwagon/precomp_multiexp.go

```diff
@@ -1,135 +1,268 @@
+	inner8Bit []*LagrangeTablePoints
 }
 
+// Equal returns true if the two PrecomputeLagrange are equal.
 func (pcl PrecomputeLagrange) Equal(other PrecomputeLagrange) bool {
-	if pcl.num_points != other.num_points {
+	if pcl.numPoints != other.numPoints {
 		return false
 	}
-
-	if len(pcl.inner) != len(other.inner) {
+	if len(pcl.inner8Bit) != len(other.inner8Bit) {
 		return false
 	}
-
-	for i := 0; i < len(pcl.inner); i++ {
-		if !pcl.inner[i].Equal(*other.inner[i]) {
+	if len(pcl.inner16Bit) != len(other.inner16Bit) {
+		return false
+	}
+	for i := 0; i < len(pcl.inner8Bit); i++ {
+		if !pcl.inner8Bit[i].Equal(*other.inner8Bit[i]) {
+			return false
+		}
+	}
+	for i := 0; i < len(pcl.inner16Bit); i++ {
+		if !pcl.inner16Bit[i].Equal(*other.inner16Bit[i]) {
 			return false
 		}
 	}
-
 	return true
 }
 
+// NewPrecomputeLagrange creates a new PrecomputeLagrange from a set of points.
 func NewPrecomputeLagrange(points []Element) *PrecomputeLagrange {
-	table := make([]*LagrangeTablePoints, len(points))
-	parallel.Execute(len(points), func(start, end int) {
-		for i := start; i < end; i++ {
-			point := points[i]
-			table[i] = newLagrangeTablePoints(point)
+	pl := &PrecomputeLagrange{numPoints: len(points)}
+
+	g, _ := errgroup.WithContext(context.Background())
+
+	// Generate 16-bit table for points[:optimized16BitIdx]
+	g.Go(func() error {
+		numPoints := len(points)
+		if numPoints > optimized16BitIdxs {
+			numPoints = optimized16BitIdxs
 		}
+		table := make([]*LagrangeTablePoints, numPoints)
+		parallel.Execute(numPoints, func(start, end int) {
+			for i := start; i < end; i++ {
+				// Each window have 1<<16 values, and we have a total of 256/16=16 windows.
+				table[i] = newLagrangeTablePoints(points[i], 256/16, 1<<16)
+			}
+		})
+		pl.inner16Bit = table
+		return nil
 	})
 
-	return &PrecomputeLagrange{
-		inner:      table,
-		num_points: len(points),
+	// Generate the 8-bit table for points[optimized16BitIdx:]
+	if len(points)-optimized16BitIdxs > 0 {
+		g.Go(func() error {
+			numPoints := len(points) - optimized16BitIdxs
+			table := make([]*LagrangeTablePoints, numPoints)
+			parallel.Execute(numPoints, func(start, end int) {
+				// We generate the table, but just shifted `optimized16BitIdxs` positions,
+				// since those group elements live in the 16-bit table.
+				for i := start; i < end; i++ {
+					// Each window have 1<<8 values, and we have a total of 256/8=32 windows.
+					table[i] = newLagrangeTablePoints(points[i+optimized16BitIdxs], 256/8, 1<<8)
+				}
+			})
+			pl.inner8Bit = table
+			return nil
+		})
 	}
+	g.Wait()
+
+	return pl
 }
 
+// SerializePrecomputedLagrange serializes a PrecomputeLagrange.
+// The format is:
+// [int64(numPoints)][int64(8bitTableCount)][8BitTable1]...[8BitTableN][int64(16bitTableCount)][16BitTable1]...[16BitTableN]
+// See (*LagrangeTablePoints).Serialize() for the format of the tables.
 func (pcl *PrecomputeLagrange) SerializePrecomputedLagrange(w io.Writer) error {
```

Comment:
Both in `SerializePrecomptuedLagrange()` and `DeserializePrecomputedLagrange()` I did some reasonable changes that I think doesn't require much comments, LMK if something seems confusing.